src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,299 @@
+    }
+
+    if (mLinkMetricsReportCallback != nullptr)
+    {
+        mLinkMetricsReportCallback(&aAddress, metrics, metricsCount, mLinkMetricsReportCallbackContext);
+    }
+
+exit:
+    return;
+}
+
+void LinkMetrics::SetLinkMetricsReportCallback(otLinkMetricsReportCallback aCallback, void *aCallbackContext)
+{
+    mLinkMetricsReportCallback        = aCallback;
+    mLinkMetricsReportCallbackContext = aCallbackContext;
+}
+
+otError LinkMetrics::SendLinkMetricsQuery(const Ip6::Address &     aDestination,
+                                          uint8_t                  aSeriesId,
+                                          const LinkMetricsTypeId *aTypeIdFlags,
+                                          uint8_t                  aTypeIdFlagsCount)
+{
+    otError                 error = OT_ERROR_NONE;
+    Tlv                     tlv;
+    LinkMetricsQueryId      linkMetricsQueryId;
+    LinkMetricsQueryOptions linkMetricsQueryOptions;
+    uint8_t                 length = 0;
+    static const uint8_t    tlvs[] = {Mle::Tlv::kLinkMetricsReport};
+    uint8_t                 buf[sizeof(Tlv) * 3 + sizeof(LinkMetricsQueryId) + sizeof(LinkMetricsQueryOptions)];
+
+    // Link Metrics Query TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsQuery);
+    length += sizeof(Tlv);
+
+    // Link Metrics Query Id sub-TLV
+    linkMetricsQueryId.Init();
+    linkMetricsQueryId.SetSeriesId(aSeriesId);
+    memcpy(buf + length, &linkMetricsQueryId, sizeof(Tlv) + linkMetricsQueryId.GetLength());
+    length += sizeof(Tlv) + linkMetricsQueryId.GetLength();
+
+    // Link Metrics Query Options sub-TLV
+    if (aTypeIdFlagsCount > 0)
+    {
+        linkMetricsQueryOptions.Init();
+        linkMetricsQueryOptions.SetLinkMetricsTypeIdList(aTypeIdFlags, aTypeIdFlagsCount);
+
+        memcpy(buf + length, &linkMetricsQueryOptions, sizeof(Tlv) + linkMetricsQueryOptions.GetLength());
+        length += sizeof(Tlv) + linkMetricsQueryOptions.GetLength();
+    }
+
+    // Set Length for Link Metrics Report Tlv
+    tlv.SetLength(length - sizeof(Tlv));
+    memcpy(buf, &tlv, sizeof(Tlv));
+
+    SuccessOrExit(error = Get<Mle::MleRouter>().SendDataRequest(aDestination, tlvs, sizeof(tlvs), 0, buf, length));
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendSingleProbeLinkMetricsReport(Message &                      aMessage,
+                                                        uint8_t &                      aLength,
+                                                        const LinkMetricsQueryOptions *aQueryOptions,
+                                                        const int8_t                   aNoiseFloor,
+                                                        const Message &                aRequestMessage)
+{
+    otError                  error = OT_ERROR_NONE;
+    LinkMetricsReportSubTlv  metric;
+    const LinkMetricsTypeId *linkMetricsTypeIds;
+    uint8_t                  metricsCount = 0;
+
+    OT_ASSERT(aQueryOptions != nullptr);
```

Comment:
A query could be as a **Single Probe** or **Forward Tracking Series**. When it's **Forward Tracking Series**, Query Options shouldn't be included in Query TLV payload.

Note that this function is only for handling **Single Probe** and it would be meaningless if no Query Options are specified.